# PED - Práctica - juego de los euromillones
#### Adrián José Zapater Reig

## Librerías requeridas:
- Pandas
- NumPy

In [1]:
import numpy as np
import pandas as pd

### Ejercicio 1:
Explora el fichero y decide cómo tienes que realizar la importación para poder guardar la información. ¿En qué estructura lo vas a guardar y qué información concreta contiene? Escribe el código asociado y trata de optimizar el espacio utilizado.

Cargamos el fichero CSV a un Dataframe para poder analizarlo y manipularlo con facilidad:

In [2]:
import os

path_to_src_file = os.path.join('source_data','historicoEuromillones.csv')
src_data = pd.read_csv(path_to_src_file)

In [3]:
src_data

FECHA  COMB. GANADORA  Unnamed: 2  Unnamed: 3  Unnamed: 4  \
0     13/10/2020               5          14          38          41   
1      9/10/2020              11          15          35          41   
2      6/10/2020               4          21          36          41   
3      2/10/2020               6          12          15          40   
4     29/09/2020               4          14          16          41   
...          ...             ...         ...         ...         ...   
1358  12/03/2004              15          24          28          44   
1359   5/03/2004               4           7          33          37   
1360  27/02/2004              14          18          19          31   
1361  20/02/2004               7          13          39          47   
1362  13/02/2004              16          29          32          36   

      Unnamed: 5  Unnamed: 6  ESTRELLAS  Unnamed: 8  
0             46         NaN          1          10  
1             50         NaN          5           8  
2             47         NaN          9          11  
3             45         NaN          3           9  
4             44         NaN         11          12  
...          ...         ...        ...         ...  
1358          47         NaN          4           5  
1359          39         NaN          1           5  
1360          37         NaN          4           5  
1361          50         NaN          2           5  
1362          41         NaN          7           9  

[1363 rows x 9 columns]

la información contenida en el CSV es:
- Fecha
- 5 columnas con la combinación ganadora
- Columna vacía con NaNs
- 2 columnas con la combinación de números estrella

### Optimización del espacio:

In [7]:
# Calculamos el espacio que ocupa el DF. Forzamos que se calcule el espacio real y no una estimación con el flag 'memory_usage=deep'
src_data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1363 entries, 0 to 1362
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   FECHA           1363 non-null   object 
 1   COMB. GANADORA  1363 non-null   int64  
 2   Unnamed: 2      1363 non-null   int64  
 3   Unnamed: 3      1363 non-null   int64  
 4   Unnamed: 4      1363 non-null   int64  
 5   Unnamed: 5      1363 non-null   int64  
 6   Unnamed: 6      0 non-null      float64
 7   ESTRELLAS       1363 non-null   int64  
 8   Unnamed: 8      1363 non-null   int64  
dtypes: float64(1), int64(7), object(1)
memory usage: 174.1 KB


El Dataframe original ocupa **174.1 KB** de memoria, y usa tipos de dato genéricos que podemos acotar porque conocemos los límites de cada columna.
Por ejemplo: los valores de los números ganadores y las estrellas van del 1-50 y 1-12 respectivamente, por lo que podemos almacenar la información en un un **uint8** y la columna 'Unnamed 6', que es un **float64** y que ocupará mucha memoria, la podemos borrar porque no contiene ningún registro.

In [12]:
# hacemos una copia del DF con la información original para no modificar el objeto src_data.
opt_src_data = src_data.copy()
# borramos la columna que no nos sirve:
opt_src_data.drop('Unnamed: 6', axis=1, inplace=True)
opt_src_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1363 entries, 0 to 1362
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   FECHA           1363 non-null   object
 1   COMB. GANADORA  1363 non-null   int64 
 2   Unnamed: 2      1363 non-null   int64 
 3   Unnamed: 3      1363 non-null   int64 
 4   Unnamed: 4      1363 non-null   int64 
 5   Unnamed: 5      1363 non-null   int64 
 6   ESTRELLAS       1363 non-null   int64 
 7   Unnamed: 8      1363 non-null   int64 
dtypes: int64(7), object(1)
memory usage: 85.3+ KB


In [ ]:
# Renombramos los indices de las columnas a
opt_src_data['']